### ISSUE TITLE
[libos] `getsockname` of `accept` returned socket gets wrong local address when listening on `INADDR_ANY`

### ISSUE THREAD

**Issue by llly** [gramineproject/gramine#725]
## Description of the problem

In the socket `accept` function, bind address of listening socket is copied to client socket as its local address.
https://github.com/gramineproject/gramine/blob/839ba4eef309350c41005e670562193a5f6d0767/libos/src/net/ip.c#L171

However when listening on `INADDR_ANY` i.e. "0.0.0.0" to receive connect from all interfaces, address of client socket is not "0.0.0.0", but a certain interface, like "127.0.0.1".

I propose to update Pal interface
```
int _PalSocketAccept(PAL_HANDLE handle, pal_stream_options_t options, PAL_HANDLE* out_client,
                     struct pal_socket_addr* out_client_addr);
```
to
```
int _PalSocketAccept(PAL_HANDLE handle, pal_stream_options_t options, PAL_HANDLE* out_client,
                     struct pal_socket_addr* out_client_addr, struct pal_socket_addr* out_local_addr);
```
and call `getsockname` just after `accept` to get local address of client socket in both linux and linux-sgx pal.  Related `ocall_accept`, `tcp_accept`, etc. interfaces are also updated.
## Steps to reproduce
Add
```
    char local_ip[INET_ADDRSTRLEN];
    uint16_t local_port;
    struct sockaddr_in local_addr;
    bzero(&local_addr, sizeof(local_addr));
    socklen_t len = sizeof(local_addr);
    if (getsockname(client_socket, (struct sockaddr*)&local_addr, &len) < 0) {
        perror("getsockname");
        exit(1);
    }
    inet_ntop(AF_INET, &local_addr.sin_addr, local_ip, sizeof(local_ip));
    local_port = ntohs(local_addr.sin_port);
    printf("[server] client_socket IP address: %s\n", local_ip);
    printf("[server] client_socket port : %u\n", local_port);
```
after
https://github.com/gramineproject/gramine/blob/839ba4eef309350c41005e670562193a5f6d0767/libos/test/regression/tcp_msg_peek.c#L78

## Expected results
```
[server] client_socket IP address: 127.0.0.1
[server] client_socket port : 11111
```
## Actual results
```
[server] client_socket IP address:  0.0.0.0
[server] client_socket port : 11111
```
## Gramine commit hash
839ba4eef309350c41005e670562193a5f6d0767

**Comment by dimakuv** [comment]
Legit issue.

@boryspoplawski What do you think about the proposal?

@llly Do you want/have time to implement the proposal (after Borys acks that this design is good)?

**Comment by llly** [comment]
Yes. I can implement it.

**Comment by dimakuv** [comment]
Great! But first let's wait for @boryspoplawski's opinion on the proposed solution.

**Comment by boryspoplawski** [comment]
Yes, the proposal looks good (bar some comments below). Out of curiosity, did you hit this on some actual software?

1. I would augment `ocall_accept` instead of calling `getsockname` not to issue more ocalls. Or did you meant that? Wasn't entirely sure, anyway this I would do.
2. It would be good to use this local address obtained from accept only if the socket was bound to `0.0.0.0`. The only place I see that could do this is `accept` in `libos/src/net/ip.c` (because PAL does not save addresses). You need to work on the details on how to do that exactly here :)
3. Please also augment some test to verify this (`tcp_msg_peek.c` you mentioned sounds good, but it's rewritten in https://github.com/gramineproject/gramine/pull/684 which we will merge soon).

**Comment by llly** [comment]
Thanks Borys. I didn't hit it on actual software. I found it by some Java netty tests.
1. That's what I want, no new ocall.
2. It make sense. Addresses can be inconsistent because that are from out of enclave.
3. Got it.
